### ISSUE TITLE
[flutter_markdown] Links do not update onTapLink when changed

### ISSUE THREAD

**Issue by creativecreatorormaybenot** [flutter/flutter#86138]
## Bug

This is a bug with the [`flutter_markdown`](https://github.com/flutter/packages/tree/master/packages/flutter_markdown/lib/src) package, where updating links will not update their `onTapLink` when the link is changed.

### Sample

When you change:

```dart
MarkdownBody(
  data: '[some link](https://foo.com)',
  onTapLink: (text, href, title) {
    print(href);
  },
),
```

To:

```dart
MarkdownBody(
  data: '[some link](https://bar.com)',
  onTapLink: (text, href, title) {
    print(href);
  },
),
```

Without reinserting the widget (so no new `initState` only re`build`) and then **click** on the link, it will still print `https://foo.com`, even though it should print bar by that point.

### Reproducing

I wanted to keep this as high-level as possible to highlight the issue (it is pretty obvious why it happens with the explanation).  
If you still wonder how to reproduce it, an example would be:

* `StatefulWidget`
* Store the `data = '[some link](https://foo.com)'` in the `State`
* Add a `GestureDetector`
* Run `setState(() { data = '[some link](https://bar.com)'; }` on tap

I did not have time to do a beautiful sample, but this shows it perfectly: no matter how often you tap the link, it will always print `foo.com`.

<details><summary>Click to expand full sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      // Application name
      title: 'Flutter Hello World',
      // Application theme data, you can set the colors for the application as
      // you want
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      // A widget which will be started on application startup
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  const MyHomePage({@required this.title});

  @override
  Widget build(BuildContext context) {
    var data = '[some link](https://foo.com)';
    return Scaffold(
      appBar: AppBar(
        // The title text which will be shown on the action bar
        title: Text(title),
      ),
      body: Center(
        child: StatefulBuilder(
          builder: (context, setState) {
            return GestureDetector(
              behavior: HitTestBehavior.opaque,
              onTap: () {
                setState(() {
                  data = '[some link](https://bar.com)';
                });
              },
              child: IgnorePointer(
                child: MarkdownBody(
                  data: data,
                  onTapLink: (_, href, __) => print(href),
                ),
              ),
            );
          },
        ),
      ),
    );
  }
}

```
</details>

### Version

This is a bug with `flutter_markdown: 0.6.2`.

## Cause

This is caused by inproper handling of rebuilding the `StatefulWidget` that handles the links :) Probably the `didUpdateWidget` in `flutter_markdown` needs to be fixed.  
Note that this does properly work when you change any of the displayed text, so e.g. change `[a link](https://example.com)` to `[another link](https://foo.com)`. But it does not work when only the link href changes (which is an actual use case we have).

**Comment by TahaTesser** [comment]
Hi @creativecreatorormaybenot 
Thanks, for filling the issue with such care, the issue is reproducible, it could be an easy fix too.

|  `stable`  | `master` |
| :---: |:---: |
|  ❌  |   ❌  | 

Check `flutter doctor -v` outputs for each channel below

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.3, on Microsoft Windows [Version 10.0.22000.65], locale en-US)
    • Flutter version 2.2.3 at C:\Users\Taha\Code\flutter_stable
    • Framework revision f4abaa0735 (8 days ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java.exe
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.19041.0

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/windows#android-setup for detailed instructions).

[✓] VS Code (version 1.58.0)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (3 available)
    • Windows (desktop) • windows • windows-x64    • Microsoft Windows [Version 10.0.22000.65]
    • Chrome (web)      • chrome  • web-javascript • Google Chrome 91.0.4472.124
    • Edge (web)        • edge    • web-javascript • Microsoft Edge 91.0.864.64

! Doctor found issues in 1 category.
```

```console
[✓] Flutter (Channel master, 2.4.0-1.0.pre.143, on Microsoft Windows [Version 10.0.22000.65], locale en-US)
    • Flutter version 2.4.0-1.0.pre.143 at C:\Users\Taha\Code\flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 311920b326 (4 hours ago), 2021-07-09 02:06:03 -0400
    • Engine revision dc06724099
    • Dart version 2.14.0 (build 2.14.0-291.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at C:\Users\Taha\Code\SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = C:\Users\Taha\Code\SDK
    • Java binary at: C:\Users\Taha\Code\android-studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)
    • All Android licenses accepted.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.10.3)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.10.31424.327
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.2.0)
    • Android Studio at C:\Users\Taha\Code\android-studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6842174)

[✓] VS Code (version 1.58.0)
    • VS Code at C:\Users\Taha\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.24.0

[✓] Connected device (4 available)
    • Windows (desktop)       • windows • windows-x64     • Microsoft Windows [Version 10.0.22000.65]
    • Windows (UWP) (desktop) • winuwp  • windows-uwp-x64 •
    • Chrome (web)            • chrome  • web-javascript  • Google Chrome 91.0.4472.124
    • Edge (web)              • edge    • web-javascript  • Microsoft Edge 91.0.864.64

• No issues found!
```
		
</details>

✅ : No Issue ❌: Issue reproduced

**Comment by abdlrhmanshehatamoussa** [comment]
Hi @TahaTesser 

I'm looking forward to participating in Flutter.
As this is my first commit, I searched for easy fix issues, until I found this one.

I will read all the contribution guides first, however I wanted to ask you, is there a way to know that there is no one else currently working on fixing this bug ?

I searched in the Pull Requests, but couldn't filter by related issues.

Thanks !

**Comment by creativecreatorormaybenot** [comment]
@abdlrhmanshehatamoussa I am pretty sure noone is working on this issue right now since it seems noone else has discovered it yet :)

**Comment by abdlrhmanshehatamoussa** [comment]
Hi Guys @creativecreatorormaybenot, @TahaTesser 

When I used the attached example, the issue was reproduced, but when I analyzed the example I found few notes that I thought it's worthy sharing with you.
I know that the example was just for the purpose of showing the issue, I'm just sharing what happened because I'm little bit in doubt about the issue, so I thought this might help reaching a better description for the issue or maybe more details:

* First, in the example there weren't any UI space left for the user to click on the `GestureDetector`, consequently the `onTap` function were never called, and the `data` variable has not changed, for the first instance it made some sense why the link of the `MarkdownBody` widget is not updated.

* So what I did, is that I inserted some extra space in the UI so that the `GestureDetector` is clickable and the function mentioned to be triggered so that the new link is updated and reflected in the `MarkdownBody`, (but still not working)

* Then, by Looking into the example code, I realized that the `data` variable is defined inside the build method of the `StatefulBuilder`, by adding some printers, it was obvious that each time the `GestureDetector` is clicked, the whole `StatefulBuilder` is being re-rendered causing the `data` variable to be redefined again with it's initial value and thus the link is never updated also.

* By moving the declaration line of the `data` variable, outside the build methods (class scope), I was sure that the value of this variable is properly changed whenever the `GestureDetector` is clicked, however the link of the `Markdown` is not yet changed ! now that sounds really like a bug !

* Lastly, I realized that changing the other value to `[another link](https://bar.com)` instead of `[some link](https://bar.com)` made it work; the new link was reflected successfully in the `MarkdownBody` widget as well as the `href` being printed properly.

**Conclusion**
The issue is better not to be generalized on all the link updates, maybe we can say it happens only when you change the link of a `MarkdownWidget` to another link having the same **name** but with different URL.

May be the historical values of the links are saved somewhere and when it happens that two links have the same name, the first one saved is always retrieved no matter what.

I will try to figure out what is happening there 

You can find my minor modifications on the example code which @creativecreatorormaybenot thankfully provided.

```
import 'package:flutter/material.dart';
import 'package:flutter_markdown/flutter_markdown.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      // Application name
      title: 'Flutter Hello World',
      // Application theme data, you can set the colors for the application as
      // you want
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      // A widget which will be started on application startup
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final String title;

  MyHomePage({required this.title});

  var data = '[some link](https://foo.com)';

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        // The title text which will be shown on the action bar
        title: Text(title),
      ),
      body: Center(
        child: StatefulBuilder(
          builder: (context, setState) {
            return GestureDetector(
              onTap: () {
                setState(() {
                  //If you change this to [same link](https://bar.com), it won't work !
                  data = '[another link](https://bar.com)';
                });
              },
              child: Column(
                children: [
                  MarkdownBody(
                    data: data,
                    onTapLink: (_, href, __) => print(href),
                  ),
                  //Extra space for triggering the onTap function of the GestureDetector
                  SizedBox(
                    height: 30,
                    child: Text("Click Me ! Current Data = $data"),
                  )
                ],
              ),
            );
          },
        ),
      ),
    );
  }
}
```

**_Note_**: Another weird behavior happens here is that you must click twice on the newly added space to trigger the function properly (may be it's something related to the `StatefulBuilder` that I'm not familiar with) 

Regards,

**Comment by creativecreatorormaybenot** [comment]
@abdlrhmanshehatamoussa Yes, I know.

What you described is what I was aware of initially, so I tried to phrase it that way in the original comment as well. So it was clear to me that the cause was the text not changing (that is what I meant with the bug in `didUpdateWidget` probably).

Anyway, I updated my original comment to hopefully make it clearer, thanks @abdlrhmanshehatamoussa ❤️   
I also updated changed two lines in the snippet to correct what you pointed out - can you check if my updated snippet works for you? And what would be really cool is if you could wrap your snippet in a `<summary>` as well 🙏

**Comment by abdlrhmanshehatamoussa** [comment]
@creativecreatorormaybenot 

I checked your snippet, sounds great, it worked with me yes, Thanks for your modifications.
I modified mine also to wrap with a summary, that's much better Thanks for your recommendation !

One minor typo in your newly added section in the description
> Note that this does properly work when you change any of the displayed text, so e.g. change `[a link](https://example.com)` to another link`. But it does not work when only the link href changes (which is an actual use case we have).

You just need to wrap the other link in code quotes, as you did with the first link.

I will try to investigate the issue and probably will start by the `didUpdateWidget` as you mentioned, Thanks for all these details : )

**Comment by salimazeriDev** [comment]
Hi, shouldn't it be closed as [[flutter_markdown] [bug-fix] [#86138] Forcing a unique key for RichText widgets](https://github.com/flutter/packages/pull/415) is merged?

**Comment by tirth-multipl** [comment]
I think this issue can be closed.
